### ISSUE TITLE
Edge indexes mismatch in NeighborSamplerWithWeights

### ISSUE THREAD

**Issue by pirakd** [bowang-lab/BIONIC#13]
I believe there is a mismatch between the edge indexing of Trainer and NeighborSamplerWithWeights. 
In Trainer, self loops appear at the end of each node's edges, but when performing 
edge_index = torch.stack([row, col], dim=0) in [line 66](https://github.com/bowang-lab/BIONIC/blob/2d3fe833034ccce3ca13598b952f113ec1c6d981/bionic/utils/sampler.py#L66)
edges appear sorted by their nodes' indexes.

Am I missing something?

**Comment by duncster94** [comment]
Hey, thanks for opening an issue. I think this is expected behavior. In the `forward` pass of `WGATConv` any self loops are removed from the existing `edge_index` Tensor and then added back in at the end of the Tensor (see [here](https://github.com/bowang-lab/BIONIC/blob/2d3fe833034ccce3ca13598b952f113ec1c6d981/bionic/model/layers.py#L86-L89)).

Does this sound right?

**Comment by pirakd** [comment]
each data flow(i.e batch) contains edge indexes in data_flows[index].Adj.e_id.
Those indexes are not aligned with Trainer.adj.[index].edge_index.  I faced this when tried to retrieve attention weights from the model.  This was not expected for me.
To fix this I replaced ([in here](https://github.com/bowang-lab/BIONIC/blob/2d3fe833034ccce3ca13598b952f113ec1c6d981/bionic/utils/sampler.py#L66))
        row, col, edge_attr = data.adj_t.t().coo()
        edge_index = torch.stack([row, col], dim=0)
with 
        edge_index = data.edge_index

So whatever suits you. Thank you for the response and for the code!

**Comment by duncster94** [comment]
Hi, thanks again for the information. I noticed that the edge weights are incorrectly set. [This](https://github.com/bowang-lab/BIONIC/blob/2d3fe833034ccce3ca13598b952f113ec1c6d981/bionic/utils/sampler.py#L65-L67) block should be this instead:

        row, col, edge_attr = data.adj_t.t().coo()
        edge_index = torch.stack([row, col], dim=0)
        self.weights = edge_attr

Can you try this change instead and let me know if it fixes your issue? Based on my testing, it looks like setting `edge_index = data.edge_index` as you suggested causes a misalignment between the input and the reconstruction target ([here](https://github.com/bowang-lab/BIONIC/blob/2d3fe833034ccce3ca13598b952f113ec1c6d981/bionic/model/loss.py#L16)). Could you also provide me with a minimal example of what attention weights you expect and what you're actually getting?

**Comment by pirakd** [comment]
I couldn't find a way to retrieve the attention weights without adding some code to BIONIC class but my problem has nothing to do with It directly.
I added the line
`edge_pairs = self.adj[0].edge_index[:, data_flows[0][2].e_id].tolist()`
after line [378](https://github.com/bowang-lab/BIONIC/blob/fb2ee913a012a99900694f7b7119ce3dba513ee6/bionic/train.py#L378) in order to retrieve the edges the were involved in the batch, but there is a mismatch between the indexes of the edges due to the reason I stated above (I think).

**Comment by duncster94** [comment]
Hi, I think I understand what's going on. The original `self.adj` datasets are only used to initialize the `NeighborSamplerWithWeights` sampler, after which they aren't used again (except for taking the `len`). The `NeighborSamplerWithWeights` initialization reorders the `edge_index`, so it ends up being different from the `self.adj` edge indices.

Can you try setting the `__init__` of `NeighborSamplerWithWeights` to:

    def __init__(self, data, *args, **kwargs):
        data = data.to("cpu")
        row, col, edge_attr = data.adj_t.t().coo()
        edge_index = torch.stack([row, col], dim=0)
        data.edge_index = edge_index
        self.weights = edge_attr
        super().__init__(edge_index, *args, **kwargs)

I think this will fix the issue and your `edge_pairs` variable should correspond to the correct edges in `self.adj[i].edge_index`.

**Comment by pirakd** [comment]
that's almost what I did at the first place, I was missing the self.weights part. 

thanks again for the much appreciated help.

**Comment by duncster94** [comment]
Great, glad I could help. I'll close the issue but feel free to reopen it if you have any other questions.
